### PR TITLE
[fix][client] fix receive dumplicated messages after seek to a timestamp in MultiTo…

### PR DESCRIPTION
**Pull request by aloyszhang**
### Motivation

For `MultiTopicConsumerImpl`, after seeking a timestamp, the consumer may receive duplicated messages.

### Modifications

Clearing the `incommingQueue` for `MultiTopicConsumerImpl` after seeking a timestamp success

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [x] `doc-not-needed` 

### Matching PR in forked repository

PR in forked repository: 
https://github.com/aloyszhang/pulsar/pull/13

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 1]
```diff
@@ -790,7 +790,10 @@ public CompletableFuture<Void> seekAsync(MessageId messageId) {
     public CompletableFuture<Void> seekAsync(long timestamp) {
         List<CompletableFuture<Void>> futures = new ArrayList<>(consumers.size());
         consumers.values().forEach(consumer -> futures.add(consumer.seekAsync(timestamp)));
-        return FutureUtil.waitForAll(futures);
+        return FutureUtil.waitForAll(futures).thenApply(f -> {
+                    clearIncomingMessages();
```
Comment (poorbarcode) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:794]: Is there a scenario like this?
- inner-consumer-1 seeks finished
- inner-consumer-2 seeks finished
- receive messages from inner-consumer-1, and push them to incoming-queue
- clear incoming-queue

then some messages are lost.

Comment (AnonHxy) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:794]: It looks that line2148 has already clear the incoming messages queue, I wonder why we also need do it again here
https://github.com/apache/pulsar/blob/af1b6e16ad9ffc0f5fad532e71c25e3a33e389c5/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L2133-L2149

Comment (aloyszhang) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:794]: I'll check this.

Comment (aloyszhang) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:794]: This only clears the incoming message queue of `ConsumerImpl`,  the `MultiTopicConsumer` also has its own  incoming message queue, messages be put into this queue from `ConsumerImpl` which may leads to duplicated messages.

Comment (aloyszhang) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:794]: You're right, there is a chance for data loss.
Change to draft first.

Comment (aloyszhang) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:828]: Update, PTAL again.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 2]
```diff
@@ -234,7 +239,21 @@ private void startReceivingMessages(List<ConsumerImpl<T>> newConsumers) {
         }
     }
 
+    private void waitForSeekComplete() {
+        while (duringSeek.get()) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] [{}] is during seek processing", topic, subscription);
+            }
+            try {
+                Thread.sleep(seekCompleteCheckTicketMs);
```
Comment (lhotari) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:248]: Thread.sleep shouldn't be run on any Netty threads. I think this solution has to be revisited. Which type of thread pool is this code running on?

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 3]
```diff
@@ -716,31 +735,46 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
...
     @Override
     public void seek(MessageId messageId) throws PulsarClientException {
+        setDuringSeek();
         try {
             seekAsync(messageId).get();
         } catch (Exception e) {
+            duringSeek.set(false);
             throw PulsarClientException.unwrap(e);
         }
+        duringSeek.set(false);
```
Comment (lhotari) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:753]: use a finally block instead

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 4]
```diff
@@ -716,31 +735,46 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
...
     }
 
     @Override
     public void seek(long timestamp) throws PulsarClientException {
+        setDuringSeek();
         try {
             seekAsync(timestamp).get();
         } catch (Exception e) {
+            duringSeek.set(false);
```
Comment (lhotari) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:762]: use a finally block instead

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 5]
```diff
@@ -716,31 +735,46 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
...
     @Override
     public void seek(Function<String, Object> function) throws PulsarClientException {
+        setDuringSeek();
         try {
             seekAsync(function).get();
         } catch (Exception e) {
+            duringSeek.set(false);
             throw PulsarClientException.unwrap(e);
         }
+        duringSeek.set(false);
```
Comment (lhotari) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:777]: use a finally block instead

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java [hunk 6]
```diff
@@ -628,6 +628,15 @@
...
     ConsumerBuilder<T> autoUpdatePartitionsInterval(int interval, TimeUnit unit);
 
+    /**
+     * Sets the interval of checking whether seek is complete interval.
+     *
+     * @param interval
+     *            the interval of check ticket.
+     * @return the consumer builder instance
+     */
+    ConsumerBuilder<T> seekCompleteCheckTicketMs(int interval);
```
Comment (lhotari) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java:638]: the name "seekCompleteCheckTicketMs" is very confusing. A better name is "seekCompletionPollIntervalMs" etc. . I wonder if the timeunit solution should be used to be consistent with other methods that configure time intervals.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java (hunk 2 above)
Comment (aloyszhang) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:248]: It's not a Netty thread. 
> Bookkeeper code includes Thread.sleep, but we should avoid that in Pulsar code base.
change to schedule later.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java (hunk 3 above)
Comment (aloyszhang) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java]: sure

**Comment by aloyszhang** [comment]
cc @lhotari

**Comment by aloyszhang** [comment]
PTAL @poorbarcode @AnonHxy

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:823]: If a task runs at this line below, we call it `receiving-task`.

https://github.com/apache/pulsar/blob/f9af4245e0b05c382656fc674fdaeda26487258c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java#L244

Then is there a scenario like this?

| time | `receiving-task` | `seek task` |
| --- | --- | --- |
| 1 | `receiving-task` started, and it is not finished | | 
| 2 |  | seek | 
| 3 |  | clear the `incoming queue` after seeking |
| 4 | `receiving-task` finished, and push messages to `incoming queue` |

Then the consumer receives duplicated messages

**Comment by poorbarcode** [comment]
Since we will start the RC version of `3.0.0` on `2023-04-11`, I will change the label/milestone of PR who have not been merged.
- The PR of type `feature` is deferred to `3.1.0`
- The PR of type `fix` is deferred to `3.0.1`

So drag this PR to `3.0.1`

**Pull request closed without merging**